### PR TITLE
Conditionally run cpp

### PR DESCRIPTION
**Pull request by shayne-fletcher**
This PR relates to discussion in issue https://github.com/ndmitchell/hlint/issues/1075.

We have been running cpp without regard for whether the `CPP` extension is enabled or not. This change avoids running cpp if `CPP` is not in effect. 

Given,
```haskell
{-# Rules
-- Foo, bar, baz
#-}
main = undefined
```
then with `hlint ~/Foo.hs` the warning
```
Warning: unknown directive #-}
in Foo.hs  at line 3 col 1
```
exhibits but with `hlint -XNoCPP ~/Foo.hs` the warning is vanquished.

### CODE REVIEW

File: src/GHC/All.hs [hunk 1]
```diff
@@ -161,17 +161,25 @@ parseModuleEx flags file str = timedIO "Parse" file $ do
...
+        dynFlags <- parsePragmasIntoDynFlags baseDynFlags enableDisableExts file str
         case dynFlags of
           Right ghcFlags -> do
             ghcFlags <- pure $ lang_set ghcFlags $ baseLanguage flags
-            case fileToModule file ppstr ghcFlags of
+            -- Avoid running cpp unless Cpp is enabled (see
+            -- https://github.com/ndmitchell/hlint/issues/1075 as to
+            -- why this is a good idea).
+            srcText <-
+              if Cpp `xopt` ghcFlags
```
Comment (ndmitchell) [inline src/GHC/All.hs:173]: There is quite a lot of vertical space here. No need for separate lines for then and else to live on their own.

File: src/GHC/All.hs [hunk 2]
```diff
@@ -161,17 +161,25 @@ parseModuleEx flags file str = timedIO "Parse" file $ do
             Nothing | file == "-" -> getContentsUTF8
                     | otherwise -> readFileUTF8' file
         str <- pure $ dropPrefix "\65279" str -- remove the BOM if it exists, see #130
-        ppstr <- runCpp (cppFlags flags) file str
         let enableDisableExts = ghcExtensionsFromParseFlags flags
-        dynFlags <- parsePragmasIntoDynFlags baseDynFlags enableDisableExts file ppstr
+        dynFlags <- parsePragmasIntoDynFlags baseDynFlags enableDisableExts file str
```
Comment (ndmitchell) [inline src/GHC/All.hs:165]: I think its important to get the pragmas after CPP. Imagine someone does:

```
{-# LANGUAGE CPP #-}
#if condition
{-# LANGUAGE QuasiQuotes #-}
#endif
```

GHC handles that.

Comment (shayne-fletcher) [reply src/GHC/All.hs:165]: Ok, good point. What this means is (1) parse pragmas, (2) if CPP is enabled, do cpp and (3) parse pragmas again. I just looked into `GHC/Driver/Pipeline.hs` and indeed that's what GHC does.

**Comment by shayne-fletcher** [comment]
In https://github.com/ndmitchell/hlint/pull/1076/commits/3f40a26769d8655b7702127d4db317284e22cc97 `parseModuleEx`re-reads pragmas after preprocessing. To help keep the resulting code tractable I've made use of an `ExcepT` monad which induces a dependency on `mtl`. Let me know if this meets with your approval or you'd rather the more verbose code that avoids the dependency? Still thinking about how to write a test for this. Haven't had any good ideas just yet.

**Comment by shayne-fletcher** [comment]
@ndmitchell Happy days - https://github.com/ndmitchell/hlint/pull/1076/commits/7c1707e9ab6cbc2eec9ab14d0a0fd4804802e389 adds a test that is confirmed to fail if `-XNoCPP` is not respected and the C-preprocessor is run.

**Comment by shayne-fletcher** [comment]
https://github.com/ndmitchell/hlint/pull/1076/commits/70a67ba36181f1cbbddb9748dc7cfbfef22b5869 adds tests that for the behaviors specified in comment https://github.com/ndmitchell/hlint/pull/1076#discussion_r452999392 (conditional extension enabling).

File: hlint.cabal [hunk 3]
```diff
@@ -69,7 +69,8 @@ library
         extra >= 1.7.3,
         refact >= 0.3,
         aeson >= 1.1.2.0,
-        filepattern >= 0.1.1
+        filepattern >= 0.1.1,
+        mtl >= 2.2.2
```
Comment (ndmitchell) [inline hlint.cabal:73]: We already depend on transformers, so no need to introduce an mtl dependency.

File: src/GHC/All.hs [hunk 4]
```diff
@@ -9,6 +9,7 @@ module GHC.All(
     parseExpGhcWithMode, parseImportDeclGhcWithMode, parseDeclGhcWithMode,
     ) where
 
+import Control.Monad.Except
```
Comment (ndmitchell) [inline src/GHC/All.hs:12]: Use Control.Monad.Trans.Except instead which is in transformers.

**Comment by ndmitchell** [comment]
Thanks! Much better

**Pull request merged**